noise-research/secure-transfer.md

```diff
@@ -0,0 +1,235 @@
+# Secure Transfers with Noise
+
+
+In this document we describe a compound protocol to enable two devices to mutually authenticate and securely exchange (arbitrary) information. 
+It consists of two main subprotocols or *phases*:
+
+- [Device Pairing](#Device-Pairing): the devices exchange and authenticate their long term device ID static keys;
+- [Secure Transfer](#Secure-Transfer): the devices securely exchange in encrypted form information using key material obtained during a successful pairing phase.
+
+## Device Pairing
+
+In the pairing phase, a device `B` requests to be paired to a device `A`. Once the two devices are paired, the devices will be mutually authenticated and will share a Noise session within which they can securely exchange information.
+
+The requests is made by exposing a QR code that, by default, has to be scanned by device `A`. 
+If device `A` doesn't have a camera while device `B` does, [it is possible](#Rationale) to execute a slightly different pairing (with same security guarantees), where `A` is exposing a QR code instead.
+
+### Employed Cryptographic Primitives
+
+- `H`: the underlying hash function, i.e. SHA-256;
+- `HKDF`: a key derivation function (based on SHA-256);
+- `Curve25519`: the underlying elliptic curve for Diffie-Hellman (DH) operations.
+
+### The `WakuPairing` Noise Handshake
+The devices execute a custom handshake derived from `X1X1`, where they mutually exchange and authenticate their device static keys, i.e.
+
+```
+WakuPairing:
+0.   <- eB              {H(sB||r), contentTopic}
+     ...
+1.   -> eA, eAeB        {H(sA||s)}   [auth_code]
+2.   <- sB, eAsB        {r}
+3.   -> sA, sAeB, sAsB  {s}
+
+{}: payload,    []: user interaction
+```
+
+
+Beside the ephemeral key, all the information embedded in the QR code should be passed to the prologue of the Noise handshake (e.g. `H(sB||r)`, `contentTopic`, etc.).
+
+
+### Protocol Flow
+1. The device `B` exposes through a QR code a Base64 serialization of:
+    - An ephemeral public key `eB`;
+    - A `contentTopic` where the information exchange will take place. `contentTopic` follows [23/WAKU2-TOPICS](https://rfc.vac.dev/spec/23/#content-topics) specifications and is then of the form `/{application-name}/{application-version}/wakunoise/1/sessions-{shard-id}/{random-message-id}/proto` for a randomly generated `random-message-id`. `contentTopic` can be then serialized in compressed form as `{application-name}:{application-version}:{shard-id}:{random-message-id}`.
```

Comment:
What is the shard id?

Reply:
> Why do we need a message id in the content topic? I'd avoid it if possible.

@s1fr0 could explain better, but afaik this is generated (or can be generated) for each message and allows for (only) the two parties to identify the next message to decrypt. Reasoning here: https://github.com/status-im/nwaku/issues/1081. However, this reminds me that we'd probably need to call it something other than `message-id`, as this concept already exists in libp2p for deduplication and is generated for each message on the GossipSub layer.